File: Android/NatStep/src/main/java/api/natsuite/natstep/StepCounter.java
/*
*   StepCounter.java
*   NatStep
*
*   Created by Yusuf on 01/14/2020.
*   Copyright (c) 2021 Yusuf Olokoba. All rights reserved.
*/

package api.natsuite.natstep;

import android.app.Application;
import android.content.Context;
import android.content.pm.PackageManager;
import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorEventListener;
import android.hardware.SensorManager;
import android.util.Log;

/**
 * Step counter.
 */
public final class StepCounter implements SensorEventListener {

    //region --Client API--

    public interface Handler {
        /**
         * Method invoked when a new step is reported.
         * @param steps Aggregate number of steps detected by the step counter.
         * @param timestamp Timestamp of the step event in nanoseconds.
         */
        void onStep (int steps, long timestamp);
    }

    /**
     * Create a step counter.
     * @param handler Handler to receive step events.
     */
    public StepCounter (Handler handler) {
        this.handler = handler;
        try {
            Application application = (Application)Class
                .forName("android.app.ActivityThread")
                .getMethod("currentApplication")
                .invoke(null, (Object[]) null);
            this.manager = (SensorManager)application.getSystemService(Context.SENSOR_SERVICE);
            final Sensor sensor = manager.getDefaultSensor(Sensor.TYPE_STEP_COUNTER);
            manager.registerListener(this, sensor, SensorManager.SENSOR_DELAY_UI);
        } catch (Exception ex) {
            Log.e("NatSuite", "NatStep Error: Failed to acquire sensor manager", ex);
        }
    }

    /**
     * Release the step counter.
     */
    public void release () {
        if (manager != null)
            manager.unregisterListener(this);
    }

    /**
     * Does the device have a pedometer sensor.
     */
    public static boolean isAvailable () {
        try {
            Application application = (Application)Class
                    .forName("android.app.ActivityThread")
                    .getMethod("currentApplication")
                    .invoke(null, (Object[]) null);
            SensorManager manager = (SensorManager)application.getSystemService(Context.SENSOR_SERVICE);
            boolean supportsSensor = application.getPackageManager().hasSystemFeature(PackageManager.FEATURE_SENSOR_STEP_COUNTER);
            boolean hasSensor = manager.getDefaultSensor(Sensor.TYPE_STEP_COUNTER) != null;
            return supportsSensor && hasSensor;
        } catch (Exception ex) {
            Log.e("NatSuite", "NatStep Error: Failed to acquire sensor manager", ex);
            return false;
        }
    }
    //endregion

    //region --Operations--

    private final Handler handler;
    private SensorManager manager;

    @Override
    public void onAccuracyChanged (Sensor sensor, int accuracy) {}

    @Override
    public void onSensorChanged (SensorEvent event) {
        handler.onStep((int)event.values[0], event.timestamp);
    }
    //endregion
}

File: Packages/com.natsuite.natstep/Runtime/StepCounter.cs
/* 
*   NatStep
*   Copyright (c) 2021 Yusuf Olokoba.
*/

namespace NatSuite.Sensors {

    using AOT;
    using System;
    using System.Runtime.InteropServices;
    using Internal;

    /// <summary>
    /// Step counter.
    /// </summary>
    public sealed class StepCounter : IDisposable {

        #region --Client API--
        /// <summary>
        /// Is step counting available on this device?
        /// </summary>
        public static bool Available => Bridge.Available();

        /// <summary>
        /// Create a step counter.
        /// </summary>
        /// <param name="handler">Delegate to receive step events.</param>
        public StepCounter (StepCountDelegate handler) {
            this.handle = GCHandle.Alloc(handler, GCHandleType.Normal);
            this.stepCounter = Bridge.CreatePedometer(OnStep, (IntPtr)handle);
        }

        /// <summary>
        /// Dispose the step counter and release resources
        /// </summary>
        public void Dispose () {
            handle.Free();
            stepCounter.Dispose();
        }
        #endregion


        #region --Operations--

        private readonly GCHandle handle;
        private readonly IntPtr stepCounter;

        [MonoPInvokeCallback(typeof(Bridge.StepHandler))]
        private static void OnStep (IntPtr context, int steps, long timestamp) => (((GCHandle)context).Target as Action<int, long>)?.Invoke(steps, timestamp);
        #endregion
    }
    
    /// <summary>
    /// Delegate invoked when a new step has been detected.
    /// </summary>
    /// <param name="steps">Aggregate number of steps detected by the step counter.</param>
    /// <param name="timestamp">Timestamo of the detected step event in nanoseconds.</param>
    public delegate void StepCountDelegate (int steps, long timestamp);
}